### PR TITLE
[FLINK-28821][table-planner] Adjust join cost for dpp query pattern which could help more plans use dpp

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

This pr is a part of FLIP-248, [https://cwiki.apache.org/confluence/display/FLINK/FLIP-248%3A+Introduce+dynamic+partition+pruning](url). In this pr, we try to adjust join cost for dpp query pattern to reorder join order to make fact table adjacent to dim table directly.  For example, for query `Select * from fact_table, sales, items, dim_table where fact_table.fact_partition_key = dim_table.dim_date_key and xxx and dim_table.price > 100` , it meets the dpp's requirements, but `fact_table` is not adjacent to `dim_table`. So we try to change the join order to make fact table adjacent to dim table directly.

## Brief change log

- Adding util class `DynamicPartitionPruningUtils`, which include methods for judging whether one join meets dpp pattern, and abstract some util methods `DynamicPartitionPruningRule` and DPP factor both used.
- In `FlinkRelMdRowCount`, adding adjust join cost logical while join meets dpp pattern.
- Adding dpp factor related plan tests in `DynamicPartitionPruningRuleTest`.

## Verifying this change

- Adding dpp factor related plan tests in `DynamicPartitionPruningRuleTest`. Include basic test and test while join key (partition key) changed in dim side.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? now don't have docs

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            // joinKeys is null means need not consider join keys.
+            if (joinKeys != null) {
+                List<RexNode> projects = ((Project) rel).getProjects();
+                ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
+                if (inputJoinKeys.isEmpty()) {
+                    dimSideFactors.isSuitableJoinKey = false;
+                }
+            }
+
+            visitDimSide(rel.getInput(0), dimSideFactors, joinKeys);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:161]: the joinKeys should be inputJoinKeys, consider the following situation: Project on Calc with field index chagned

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                    dimSideFactors.isSuitableJoinKey = false;
+                }
+            }
+
+            RexProgram origProgram = ((Calc) rel).getProgram();
+            if (origProgram.getCondition() != null
+                    && isSuitableFilter(origProgram.expandLocalRef(origProgram.getCondition()))) {
+                dimSideFactors.hasFilter = true;
+            }
+            visitDimSide(rel.getInput(0), dimSideFactors, joinKeys);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:181]: ditto

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+import java.util.stream.Collectors;
+
+/** Planner utils for Dynamic partition Pruning. */
+public class DynamicPartitionPruningUtils {
+
+    /**
+     * For the input join node, judge whether the join left side and join right side meet the
+     * requirements of dynamic partition pruning. If meets the requirements,return true.
+     */
+    public static boolean supportDynamicPartitionPruning(Join join) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:58]: Can DynamicPartitionPruningRule#doMatches be replaced with this method ?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala [hunk 4]
```diff
@@ -335,10 +336,19 @@ class FlinkRelMdRowCount private extends MetadataHandler[BuiltInMetadata.RowCoun
       fmq.getSelectivity(joinWithOnlyEquiPred, nonEquiPred)
     }
 
+    /** Judge */
+    val dynamicPartitionPruningFactor =
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala:340]: add some comments to explain why we use factor here

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left some comments

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala (hunk 4 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala:340]: > add some comments to explain why we use factor here

done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 3 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:58]: > Can DynamicPartitionPruningRule#doMatches be replaced with this method ?

These two method have different inputs, I think it's better to reuse some methods like `isDimSide()` etc. in this method that reuse this method itself.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        return false;
+    }
+
+    /**
+     * Judge whether input RelNode meets the conditions of dimSide. If joinKeys is null means we
+     * need not consider the join keys in dim side, which already deal by dynamic partition pruning
+     * rule. If joinKeys not null means we need to judge whether joinKeys changed in dim side, if
+     * changed, this RelNode is not dim side.
+     */
+    public static boolean isDimSide(RelNode rel, @Nullable ImmutableIntList joinKeys) {
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:111]: > private
> 
> joinKeys should not be null in current situation
This method is one reused method,  If joinKeys is null means we need not consider the join keys in dim side, which already deal by dynamic partition pruning rule.

**Comment by swuferhong** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+        DppDimSideFactors dimSideFactors = new DppDimSideFactors();
+        visitDimSide(rel, dimSideFactors, joinKeys);
+        return dimSideFactors.isDimSide();
+    }
+
+    /**
+     * Visit dim side to judge whether dim side has filter condition and whether dim side's source
+     * table scan is non partitioned scan.
+     */
+    public static void visitDimSide(
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java]: > private

done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 1 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java]: > the joinKeys should be inputJoinKeys, consider the following situation: Project on Calc with field index chagned

done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 3 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:61]: > Can DynamicPartitionPruningRule#doMatches be replaced with this method ?

Already replaced.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java]: > private
> 
> joinKeys should not be null in current situation

Already modify.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 2 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java]: > ditto

Already modified.

Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java]: > ditto

Already modified.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala [hunk 7]
```diff
@@ -335,10 +336,21 @@ class FlinkRelMdRowCount private extends MetadataHandler[BuiltInMetadata.RowCoun
       fmq.getSelectivity(joinWithOnlyEquiPred, nonEquiPred)
     }
 
+    // Dynamic partition pruning factor is adding to adjust join cost for these join
+    // node which meets dynamic partition pruning pattern. Try best to reorder the
+    // fact table and fact table together to make dpp succeed.
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala:341]: Currently, join-reorder is before dynamic partition pruning rewrite. This factor is adding to adjust join cost for these join node which meets dynamic partition pruning pattern. Try best to reorder the fact table and fact table together to make DPP succeed.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+            factSideFactors.isSuitableFactScanSource = !acceptedFields.isEmpty();
+        } else if (rel instanceof HepRelVertex) {
+            visitFactSide(((HepRelVertex) rel).getCurrentRel(), factSideFactors, joinKeys);
+        } else if (rel instanceof Exchange || rel instanceof Filter) {
+            visitFactSide(rel.getInput(0), factSideFactors, joinKeys);
+        } else if (rel instanceof Project) {
+            List<RexNode> projects = ((Project) rel).getProjects();
+            ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
+            if (inputJoinKeys.isEmpty()) {
+                factSideFactors.isSuitableJoinKey = false;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:200]: return directly

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+        } else if (rel instanceof Calc) {
+            Calc calc = (Calc) rel;
+            RexProgram program = calc.getProgram();
+            List<RexNode> projects =
+                    program.getProjectList().stream()
+                            .map(program::expandLocalRef)
+                            .collect(Collectors.toList());
+            ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
+            if (inputJoinKeys.isEmpty()) {
+                factSideFactors.isSuitableJoinKey = false;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:213]: ditto

**Review by godfreyhe — COMMENTED** [review]
LGTM overall, except some minor comments

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala (hunk 7 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdRowCount.scala]: > Currently, join-reorder is before dynamic partition pruning rewrite. This factor is adding to adjust join cost for these join node which meets dynamic partition pruning pattern. Try best to reorder the fact table and fact table together to make DPP succeed.

done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 8 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:200]: > return directly

done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 9 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:214]: > ditto

done!

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
The private CI is green: https://dev.azure.com/leonardBang/Azure_CI/_build/results?buildId=715&view=results. I will merge it

**Pull request closed without merging**